pkg/server/auth.go

```diff
@@ -161,3 +165,218 @@ func (s *Server) Logout(c *gin.Context) {
+		c.JSON(http.StatusBadRequest, api.Error(errors.ErrBindJSON))
+		return
+	}
+
+	// Validate the request
+	if err = in.Validate(); err != nil {
+		c.Error(err)
+		c.JSON(http.StatusBadRequest, api.Error(err))
+		return
+	}
+
+	// Retrieve the API key from the database
+	ctx = c.Request.Context()
+	if apiKey, err = s.store.RetrieveAPIKey(ctx, in.ClientID); err != nil {
+		if errors.Is(err, errors.ErrNotFound) {
+			c.JSON(http.StatusUnauthorized, api.Error(errors.ErrFailedAuthentication))
+			return
+		}
+		c.Error(err)
+		c.JSON(http.StatusInternalServerError, api.Error(errors.ErrInternal))
+		return
+	}
+
+	// Verify the API key's secret
+	var verified bool
+	if verified, err = passwords.VerifyDerivedKey(apiKey.Secret, in.ClientSecret); err != nil {
+		c.Error(err)
+		c.JSON(http.StatusInternalServerError, api.Error(errors.ErrInternal))
+		return
+	}
+
+	if !verified {
+		c.JSON(http.StatusUnauthorized, api.Error(errors.ErrFailedAuthentication))
+		return
+	}
+
+	if err = s.store.UpdateLastSeen(ctx, apiKey.ID, time.Now()); err != nil {
+		c.Error(err)
+		c.JSON(http.StatusInternalServerError, api.Error(errors.ErrInternal))
+		return
+	}
+
+	// Prepare the login reply now that the user has been authenticated
+	out = &api.LoginReply{}
+	if apiKey.LastSeen.Valid {
+		out.LastLogin = apiKey.LastSeen.Time
+	}
+
+	// Create access and refresh tokens for the API key
+	claims = apiKey.Claims()
+	if out.AccessToken, out.RefreshToken, err = s.issuer.CreateTokens(claims); err != nil {
+		c.Error(err)
+		c.JSON(http.StatusInternalServerError, api.Error(errors.ErrInternal))
+		return
+	}
+
+	// Set cookies for the client
+	if err = auth.SetAuthCookies(c, out.AccessToken, out.RefreshToken); err != nil {
+		c.Error(err)
+		c.JSON(http.StatusInternalServerError, api.Error(errors.ErrInternal))
+		return
+	}
+
+	// Content negotiation and redirection if required.
+	switch c.NegotiateFormat(binding.MIMEJSON, binding.MIMEHTML) {
+	case binding.MIMEJSON:
+		c.JSON(http.StatusOK, out)
+	case binding.MIMEHTML:
+		htmx.Redirect(c, http.StatusSeeOther, in.Redirect())
+	default:
+		c.AbortWithError(http.StatusNotAcceptable, errors.ErrNotAccepted)
+	}
+}
+
+// Reauthenticate a user via their refresh token.
+func (s *Server) Reauthenticate(c *gin.Context) {
+	var (
+		err    error
+		claims *gimlet.Claims
+		sub    gimlet.SubjectType
+		subID  ulid.ULID
+		in     *api.ReauthenticateRequest
+		out    *api.LoginReply
+	)
+
+	if err = c.BindJSON(&in); err != nil {
+		c.Error(err)
+		c.JSON(http.StatusBadRequest, api.Error(errors.ErrBindJSON))
+		return
+	}
+
+	// Validate the request
+	if err = in.Validate(); err != nil {
+		c.Error(err)
+		c.JSON(http.StatusBadRequest, api.Error(err))
+		return
+	}
+
+	// Verify the refresh token
+	if claims, err = s.issuer.Verify(in.RefreshToken); err != nil {
+		c.Error(err)
+		c.JSON(http.StatusForbidden, api.Error(errors.ErrFailedAuthentication))
+		return
+	}
+
+	// TODO: Verify that this is a refresh token with the correct audience.
```

Comment:
Does this functionality exist in gimlet?

Reply:
It exists in the `auth` package via the issuer. I'll make the required update.